megatron/core/inference/engines/dynamic_engine.py

```diff
@@ -841,68 +854,149 @@ async def async_step(
                     mem["allocation.all.current"],
                     mem["allocated_bytes.all.current"] / (1024**3),
                     mem["reserved_bytes.all.current"] / (1024**3),
                 )
             )
-            if prev_is_decode_only:
+            if is_decode_only:
                 output_str = f"\033[94m{output_str}\033[0m"
             logging.info(output_str)
 
-        self.step_count += 1
-
-        range_pop()
         return {
             "active_requests": active_requests,
             "finished_requests": finished_requests,
             "step_time": step_time,
             "cuda_graph_request_count": cuda_graph_request_count,
         }
 
+    async def async_forward(self) -> Tuple[Dict, Tuple, float]:
+        """Uses `asyncio` for continuous generation.
+        Sleeps when no requests are available, until new requests have been added.
+
+        Returns:
+            A tuple comprised of:
+                step_result (Optional[Dict]): The result of the step.
+                context_state (Tuple): A tuple consisting of the state of the context.
+                is_decode_only, total/paused request count, active token count.
+                step_time (float): How long this step took.
+        """
+        # schedule requests
+        self.schedule_waiting_requests()
+
+        # Saving pre-step state, for printing output below.
+        is_decode_only = self.context.is_decode_only()
+        pre_step_context_state = (
+            is_decode_only,
+            self.context.total_request_count,
+            self.context.paused_request_count,
+            self.context.active_token_count,
+        )
+
+        # Generate tokens.
+        range_push("Prefill" if not is_decode_only else "Decode")
+
+        self.step_start_event.record()
+        result = await self.controller.async_generate_output_tokens_dynamic_batch()
+        self.step_end_event.record()
+        self.step_end_event.synchronize()
+        step_time = self.step_start_event.elapsed_time(self.step_end_event) / 1e3
+        self.step_count += 1
+
+        range_pop()
+
+        if (
+            self.inference_logging_step_interval > 0
+            and step_count > 0
+            and step_count % self.inference_logging_step_interval == 0
+            and self.context.metrics_writer is not None
+        ):
+            kvcache_util_stats = self.context.get_kvcache_utilization_stats()
+        else:
+            kvcache_util_stats = None
+
+        post_step_context_state = (
+            len(self.waiting_request_ids),
+            self.finished_request_count,
+            kvcache_util_stats,
+            self.context.padded_active_token_count,
+            self.context.using_cuda_graph_this_step(),
+            self.context.block_allocator.active_count,
+            self.context.block_allocator.paused_count,
+            self.context.block_allocator.get_active_used(),
+            self.context.block_allocator.get_paused_used(),
```

Comment:
@lmcafee-nvidia could you please confirm whether these values are indeed intended to be scraped after the step, as opposed to before the step?

I copied their place as-is from `main`, but I want to make sure `main` itself was correct.